artifactory/services/upload.go

```diff
@@ -273,58 +273,73 @@ func CollectFilesForUpload(uploadParams UploadParams, progressMgr ioutils.Progre
-	// 'foldersPaths' is a subset of the 'paths' array. foldersPaths is in use only when we need to upload folders with flat=true.
-	// 'foldersPaths' will contain only the directories paths which are in the 'paths' array.
-	var foldersPaths []string
-	for index, path := range paths {
-		matches, isDir, isSymlinkFlow, err := fspatterns.PrepareAndFilterPaths(path, excludePathPattern, uploadParams.IsSymlink(), uploadParams.IsIncludeDirs(), patternRegex)
+	if err != nil {
+		return err
+	}
+	var uploadedTargets []string
+	// 'uploadedDirs' is in use only when we need to upload folders with flat=true.
+	// 'uploadedDirs' will contain only local directories paths that have been uploaded to Artifactory.
+	var uploadedDirs []string
+	for _, path := range paths {
+		matches, isDir, err := fspatterns.SearchPatterns(path, uploadParams.IsSymlink(), uploadParams.IsIncludeDirs(), patternRegex)
 		if err != nil {
 			return err
 		}
-
 		if len(matches) > 0 {
-			target := uploadParams.GetTarget()
-			tempPaths := paths
-			tempIndex := index
-			// In case we need to upload directories with flat=true, we want to avoid the creation of unnecessary paths in Artifactory.
-			// To achieve this, we need to take into consideration the directories which had already been uploaded, ignoring all files paths.
-			// When flat=false we take into consideration folder paths which were created implicitly by file upload
-			if uploadParams.IsFlat() && uploadParams.IsIncludeDirs() && isDir {
-				foldersPaths = append(foldersPaths, path)
-				tempPaths = foldersPaths
-				tempIndex = len(foldersPaths) - 1
+			taskData, err := NewUploadTaskData(path, isDir, matches, len(matches), uploadParams, vcsCache)
+			if err != nil {
+				return err
 			}
-			taskData := &uploadTaskData{target: target, path: path, isDir: isDir, isSymlinkFlow: isSymlinkFlow,
-				paths: tempPaths, groups: matches, index: tempIndex, size: len(matches), uploadParams: uploadParams,
-				vcsCache: vcsCache,
+			if isDir {
+				if skipDirUpload(uploadedTargets, uploadedDirs, taskData.target, path, uploadParams.IsIncludeDirs()) {
+					continue
+				}
+				uploadedDirs = append(uploadedDirs, path)
 			}
+			// Update progress
 			incGeneralProgressTotal(progressMgr, uploadParams)
+			// Create upload task
 			err = createUploadTask(taskData, dataHandlerFunc)
 			if err != nil {
 				return err
 			}
+			uploadedTargets = append(uploadedTargets, taskData.target)
 		}
 	}
 	return nil
 }
 
+// targetFiles - Paths in Artifactory of the the files that were uploaded.
+// sourceDirs - Paths of the the local dirs that have been already uploaded to Artifactory. (Longest files path first)
+// targetDir - The directory target path to be uploaded.
+// sourceDir - The directory source path to be uploaded.
+func skipDirUpload(targetFiles, sourceDirs []string, targetDir, sourceDir string, includeDirs bool) bool {
+	// Check that the dir is not already created in Artifactory by implicitly upload child file
```

Comment:
...by implicitly upload child file --> ...following an implicit upload of child file.